tests/integration/terraform_setup.py

```diff
@@ -0,0 +1,212 @@
+  (default: the script's directory).
+- `--manifest`: Path to the YAML manifest file used by Terraform
+  (default: `default-manifest.yaml` in the script's directory).
+- `--model-name`: The name of the Juju model to operate on
+  (default: `my-canonical-k8s`).
+
+Usage:
+    python3 script.py
+"""
+
+import argparse
+import os
+import subprocess
+import sys
+from pathlib import Path
+
+
+def run_command(command: str, capture_output: bool = False, fail_on_error: bool = True) -> Optional[str]:
+    """Run a shell command.
+
+    Args:
+        command: The shell command to run.
+        capture_output: Whether to capture the command's output.
+        fail_on_error: Whether to exit the script if the command fails.
+
+    Returns:
+        The command's output if `capture_output` is `True`, otherwise `None`.
+    """
+    try:
+        result = subprocess.run(
+            command, shell=True, check=True, text=True, capture_output=capture_output
+        )
+        return result.stdout.strip() if capture_output else None
+    except subprocess.CalledProcessError as e:
+        if fail_on_error:
+            print(f"Error running command: {command}\n{e}")
+            sys.exit(1)
+        return None
+
+
+def ensure_terraform(expected_version: str) -> None:
+    """Ensure Terraform is installed and matches the expected version.
+
+    Args:
+        expected_version: The expected version of Terraform.
+    """
+    installed_version = run_command(
+        "snap list terraform | awk '/^terraform/ {print $4}'", capture_output=True
+    )
+    if not installed_version:
+        print(f"Terraform is not installed. Installing version {expected_version}...")
+        run_command(f"sudo snap install terraform --channel={expected_version} --classic")
+    elif installed_version != expected_version:
+        print(
+            f"Error: Installed Terraform version ({installed_version})"
+            f"does not match the expected version ({expected_version})."
+        )
+        sys.exit(1)
+    else:
+        print(
+            f"Terraform is already installed and matches the expected version: {installed_version}."
+        )
+
+
+def setup_juju_provider_authentication() -> None:
+    """Set up Juju provider authentication.
+
+    See https://registry.terraform.io/providers/juju/juju/latest/docs#authentication
+    """
+    controller = run_command("juju whoami | yq .Controller", capture_output=True) or ""
+    os.environ["CONTROLLER"] = controller or ""
+    os.environ["JUJU_CONTROLLER_ADDRESSES"] = (
+        run_command(
+            f"juju show-controller | yq .{controller}.details.api-endpoints"
+            "| yq -r '. | join(\",\")'",
+            capture_output=True,
+        )
+        or ""
+    )
+    os.environ["JUJU_USERNAME"] = (
+        run_command(
+            f"cat ~/.local/share/juju/accounts.yaml"
+            f"| yq .controllers.{controller}.user | tr -d '\"'",
+            capture_output=True,
```

Comment:
Small suggestion regarding `yq`. It should enable the `-r` flag by default, so you won't need the `tr` part.
```suggestion
            f"yq .controllers.{controller}.user < ~/.local/share/juju/accounts.yaml",
            capture_output=True,
```

Reply:
Can we use `yaml`?

```python
accounts = yaml.safe_load(Path("~/.local/share/juju/accounts.yaml").open())
os.envrion["JUJU_USERNAME"] = accounts["controllers"][controller]["user"].strip()
```

Reply:
I agree with @addyess. It would be better to just offload everything to Python constructs.

Reply:
I ended up using the Juju Python clientlib